runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/content/ByteStreamJVM.kt

```diff
@@ -114,11 +114,25 @@ public fun ByteStream.Companion.fromInputStream(
  * @param contentLength If specified, indicates how many bytes remain in this stream. Defaults to `null`.
  */
 public fun InputStream.asByteStream(contentLength: Long? = null): ByteStream.SourceStream {
-    val source = source()
+    if (markSupported() && contentLength != null) {
+        mark(contentLength.toInt())
+    }
+
     return object : ByteStream.SourceStream() {
         override val contentLength: Long? = contentLength
         override val isOneShot: Boolean = !markSupported()
-        override fun readFrom(): SdkSource = source
+        override fun readFrom(): SdkSource {
+            if (markSupported() && contentLength != null) {
+                reset()
+                mark(contentLength.toInt())
+                return object : SdkSource by source() {
+                    // no-op close
+                    override fun close() { }
+                }
+            }
+
+            return source()
+        }
     }
 }
```

Comment:
**Question**: Is this a behavioral change wrt closing? It seems like before we returned the raw underlying `SdkSource` before which, if closed, would close the underlying `InputStream`. Now it looks like that won't happen...why is that necessary?

Reply:
Yes it is. Without this change the underlying `InputStream` is closed after hashing (because we call a `.use { }` on the `SdkSource`) and subsequent attempts to read the body (e.g. during transmission) fail with `IOException: Stream closed`